### PR TITLE
Add glibc-2.28 (OpenSSL 1.1) wheels alongside 2.34

### PR DESCRIPTION
**Pull request by gargsaumya**
## Summary

Ship `manylinux_2_28` (OpenSSL 1.1) wheels alongside the existing `manylinux_2_34` (OpenSSL 3) wheels, extending support to older enterprise Linux distributions without dropping modern-distro support or vendoring OpenSSL.

| Host family | Preferred wheel | System OpenSSL |
|---|---|---|
| glibc 2.28╬ô├▓┬╝Γö£Γöñ╬ô├╢┬úΓö£┬║╬ô├╢┬ú╬ô├╢├▒2.33 (for example RHEL 8, Ubuntu 20.04, Debian 11) | `manylinux_2_28` | `libssl.so.1.1` / `libcrypto.so.1.1` |
| glibc 2.34+ (for example RHEL 9, Ubuntu 22.04+, Debian 12+) | `manylinux_2_34` | `libssl.so.3` / `libcrypto.so.3` |

PR #532, which supplied the manylinux repair and release infrastructure used here, is merged into `main`. This branch is reconciled with current `main`.

## Related work item

AB#47953

## Changes

- Add maturin to both `manylinux_2_28_*_rust` builder images.
- Build glibc 2.28 wheels independently from glibc 2.34 outputs, with an isolated Cargo target directory.
- Build and inject the ODBC driver in the matching libc/OpenSSL environment.
- Parameterize and enforce the maximum GLIBC symbol version (`2.28` or `2.34`).
- Repair each glibc family with the matching manylinux tag and OpenSSL exclusions.
- Explicitly pull mutable `manylinux_2_28` builder tags before first use on persistent x64 and ARM64 agents.
- Expand the exact release contract and documentation from 34 to 44 wheels.
- Add regression coverage for image refresh ordering, ODBC-disabled pipeline expansion, GLIBC ceilings, OpenSSL 1.1 repair exclusions, and the expanded wheel matrix.

## Final-head validation

Artifact-producing source: `70795fc40a2cc3dcdf20b800fe24914abf955fd7`  
Current source: `3838b6cd841af497524f390413858131457924fa` (testability and execution-level GLIBC regression coverage only)

- [ADO producer run 174990](https://dev.azure.com/sqlclientdrivers/mssql-rs/_build/results?buildId=174990) completed successfully on the exact source above.
- Both 2.28 jobs downloaded newer builder images, confirmed maturin was available, and built CPython 3.10╬ô├▓┬╝Γö£Γöñ╬ô├╢┬úΓö£┬║╬ô├╢┬ú╬ô├╢├▒3.14 wheels.
- Production wheel validator passed with bundled ODBC required: **44 wheels** exactly:
  - 10 `manylinux_2_28`
  - 10 `manylinux_2_34`
  - 10 `musllinux_1_2`
  - 5 macOS universal2
  - 5 Windows x64
  - 4 Windows ARM64
- LLVM inspection of representative CPython 3.13 x64 and ARM64 binaries proved:
  - no wheel bundles `libssl` or `libcrypto`;
  - 2.28 extension and ODBC binaries require OpenSSL 1.1 and have maximum GLIBC 2.28;
  - 2.34 extension and ODBC binaries require OpenSSL 3 and have maximum GLIBC 2.34.
- The run produced and validated both Rust release crates:
  - `mssql-tds 0.1.0-dev.20260914.174990`
  - `mssql-mock-tds 0.1.0-dev.20260914.174990`
  - `cargo package` verification compiled both packaged crates, including verification of packaged `mssql-mock-tds` against packaged `mssql-tds`.
- Published dev NuGet `mssql-python-rs-wheels 0.1.0-dev.20260914.174990` is visible on the public feed. Its ZIP central directory contains the exact same 44-wheel family counts (501,529,225 bytes total). Unmodified current `mssql-python` `origin/main` PowerShell consumer logic downloaded it, selected the CPython 3.13 Windows x64 wheel, extracted the native extension plus `mssqlodbc.dll`, and imported `mssql_py_core` successfully from an isolated directory.
- Focused release-pipeline suite: **176 passed**, with 3 known Windows ANSI/rendering assertions deselected.
- `cargo bfmt`: passed.
- `cargo bclippy`: passed in an isolated source snapshot with a freshly generated offline lockfile.
- Changed shell scripts pass `bash -n`; changed YAML parses; `git diff --check` passes.
- All 9 Copilot review threads were adjudicated as valid, fixed, replied to, and resolved. An independent final diff audit found no PR-blocking issues.

## Known validation boundary / follow-up

- `cargo btest` is **not claimed green** locally: without `DB_CONNECTION_STRING` or the certificate fixtures, 3,807 of 4,212 tests passed and 405 integration/fixture-dependent tests failed. The producer pipeline for this exact head is green.
- The current `mssql-python` Bash consumer recognizes the renamed NuGet package and wheel distribution, but its glibc selector still requests legacy bare `linux_*` tags rather than `manylinux_2_28`/`manylinux_2_34`. That downstream selector migration is a separate `mssql-python` change and is not bundled into this producer PR.
- No ESRP/PyPI/crates.io publication was attempted. Real publication remains limited to an official release after this change reaches the stable Official Build path.

## Checklist

- [x] `cargo bfmt` passes
- [x] `cargo bclippy` passes
- [ ] `cargo btest` passes locally (live SQL Server and certificate fixtures unavailable; result documented above)
- [x] New/changed functionality has tests
- [x] Operational documentation reflects the 44-wheel contract
- [x] Artifact-producing head validation is green; latest test-only head checks are tracked by the PR
- [x] Author self-review completed

### CODE REVIEW

File: .pipeline/OneBranch/stages.yml [hunk 1]
```diff
@@ -314,6 +314,51 @@ stages:
...
+        displayName: 'Build glibc-2.28 wheels (Linux x64)'
+        condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))
+        env:
+          CARGO_REGISTRIES_MSSQL_RS_PUBLIC_TOKEN: $(CARGO_REGISTRIES_MSSQL_RS_PUBLIC_TOKEN)
+          CARGO_REGISTRIES_MSSQL_RS_TOKEN: $(CARGO_REGISTRIES_MSSQL_RS_TOKEN)
+      - template: /.pipeline/templates/build-odbc-driver-in-wheel-image-template.yml
+        parameters:
+          image: 'ghcr.io/microsoft/mssql-rs/python-build/manylinux_2_28_x86_64_rust:latest'
+          libc: glibc
+          displaySuffix: 'glibc-2.28 Linux x64'
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply .pipeline/OneBranch/stages.yml]: Valid. Fixed in bb623d2f: the shared template now accepts maxGlibcVersion (default 2.34), both manylinux 2.28 calls pass 2.28, and the driver guard compares its maximum required symbol against that ceiling. Verified the guard accepts GLIBC_2.28, rejects GLIBC_2.29 for a 2.28 build, and retains the 2.34 default.

File: scripts/repair-glibc-wheels-in-container.sh [hunk 2]
```diff
@@ -1,18 +1,28 @@
...
+# on the target distro. The floor is set by the build image and passed here:
+#   manylinux_2_34 (AlmaLinux 9, OpenSSL 3)   -> exclude libssl.so.3 / libcrypto.so.3
+#   manylinux_2_28 (AlmaLinux 8, OpenSSL 1.1) -> exclude libssl.so.1.1 / libcrypto.so.1.1
+# musllinux and non-linux wheels are left untouched. Run inside the matching
+# manylinux build image (has auditwheel and the matching OpenSSL runtime).
 set -euo pipefail
 
 WHEELS_DIR="${1:?usage: repair-glibc-wheels-in-container.sh <wheels-dir>}"
 PLAT="${REPAIR_PLAT:-manylinux_2_34}"
+EXCLUDE_LIBS="${REPAIR_EXCLUDE_LIBS:-libssl.so.3 libcrypto.so.3}"
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply scripts/repair-glibc-wheels-in-container.sh:17]: Valid. Added direct regression coverage in bb623d2f. The test sets REPAIR_PLAT=manylinux_2_28 and the OpenSSL 1.1 REPAIR_EXCLUDE_LIBS override, then verifies both 2.28 output tags, both .so.1.1 exclusion arguments, and absence of the default .so.3 exclusions.

File: .pipeline/scripts/verify-python-wheels.ps1 [hunk 3]
```diff
@@ -79,6 +79,8 @@ function Get-ExpectedWheelNames {
         'win_amd64',
         'manylinux_2_34_x86_64',
         'manylinux_2_34_aarch64',
+        'manylinux_2_28_x86_64',
+        'manylinux_2_28_aarch64',
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply .pipeline/scripts/verify-python-wheels.ps1:83]: Valid. Updated the release-management architecture, matrix, nightly/dev flows, and package example to 44 wheels, including both 2.28 rows. The shared-stage header now also documents 44 wheels across all nine platform tags. Fixed in bb623d2f.

File: .pipeline/OneBranch/stages.yml [hunk 4]
```diff
@@ -314,6 +314,52 @@ stages:
...
+          bash .pipeline/scripts/docker-cargo-run.sh --rm \
+            -v "$(Build.SourcesDirectory):/workspace" \
+            -v "$(Build.SourcesDirectory)/wheels-228:/workspace/target/wheels" \
+            -e "WORKSPACE_DIR=/workspace" \
+            -e "OUTPUT_DIR=/workspace/target/wheels" \
+            -e "CARGO_TARGET_DIR=/tmp/mssql-py-core-manylinux-2-28" \
+            ghcr.io/microsoft/mssql-rs/python-build/manylinux_2_28_x86_64_rust:latest \
+            bash /workspace/scripts/build-python-wheels-in-container.sh
+        displayName: 'Build glibc-2.28 wheels (Linux x64)'
+        condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply .pipeline/OneBranch/stages.yml]: Valid. Fixed in a07528d8 after synchronizing PR #557 with its latest stacked base. For both x64 and ARM64, raw 2.28 wheel build and repair/copy now run whenever buildPythonWheels is enabled; only the 2.28 ODBC driver build and injection remain under buildOdbcNative. The expansion regression covers ODBC true/false on both architectures and verifies build -> optional injection -> repair ordering. The no-ODBC expansion contains only Build and Repair for each 2.28 family. Validation: 150 passed, 3 known Windows ANSI-rendering assertions deselected.

File: .pipeline/OneBranch/stages.yml [hunk 5]
```diff
@@ -270,6 +270,23 @@ stages:
           osType: Alpine
           architecture: x64
           publishArtifacts: false
+      # --- glibc-2.28 wheels (OpenSSL 1.1) for RHEL 8 / Ubuntu 20.04 ---
+      - script: |
+          set -e
+          mkdir -p "$(Build.SourcesDirectory)/wheels-228"
+          bash .pipeline/scripts/docker-cargo-run.sh --rm \
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply .pipeline/OneBranch/stages.yml:278]: Fixed in 70795fc4. The x64 glibc-2.28 build now explicitly pulls the mutable builder tag before its first use, so persistent agents cannot reuse the pre-rebuild image. The later ODBC build in the same job uses that refreshed local image. Added structural coverage for pull-before-run ordering.

File: .pipeline/OneBranch/stages.yml [hunk 6]
```diff
@@ -362,6 +413,23 @@ stages:
             osType: Alpine
             architecture: arm64
             publishArtifacts: false
+        # --- glibc-2.28 wheels (OpenSSL 1.1) for RHEL 8 / Ubuntu 20.04 ---
+        - script: |
+            set -e
+            mkdir -p "$(Build.SourcesDirectory)/wheels-228"
+            bash .pipeline/scripts/docker-cargo-run.sh --rm \
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply .pipeline/OneBranch/stages.yml:422]: Fixed in 70795fc4. The ARM64 glibc-2.28 build now explicitly pulls the mutable builder tag before its first use, with the same pull-before-run regression assertion.

File: .pipeline/templates/build-odbc-driver-in-wheel-image-template.yml [hunk 7]
```diff
@@ -14,6 +14,10 @@ parameters:
   # Human-readable suffix for the step display name, e.g. 'glibc Linux x64'.
   - name: displaySuffix
     type: string
+  # Maximum allowed glibc symbol version for glibc builds.
+  - name: maxGlibcVersion
+    type: string
+    default: '2.34'
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply .pipeline/templates/build-odbc-driver-in-wheel-image-template.yml:20]: Fixed in 70795fc4. The template header now documents manylinux_2_28, manylinux_2_34, and musllinux_1_2.

File: scripts/repair-glibc-wheels-in-container.sh [hunk 8]
```diff
@@ -1,18 +1,28 @@
...
-# would break server-certificate validation on Debian/Ubuntu. musllinux and
-# non-linux wheels are left untouched. Run inside the matching manylinux build
-# image (has auditwheel and the OpenSSL 3 runtime).
+# manylinux_*, which PyPI accepts. The OS-provided OpenSSL is excluded from the
+# graft (REPAIR_EXCLUDE_LIBS) so the wheels keep using the system OpenSSL:
+# native-tls resolves CA paths from the loaded libcrypto's compiled-in OPENSSLDIR,
+# so vendoring the build image's OpenSSL would break server-certificate validation
+# on the target distro. The floor is set by the build image and passed here:
+#   manylinux_2_34 (AlmaLinux 9, OpenSSL 3)   -> exclude libssl.so.3 / libcrypto.so.3
+#   manylinux_2_28 (AlmaLinux 8, OpenSSL 1.1) -> exclude libssl.so.1.1 / libcrypto.so.1.1
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply scripts/repair-glibc-wheels-in-container.sh:10]: Fixed in 70795fc4. The packaging comment now documents the OS-provided OpenSSL 1.1 expectation for manylinux_2_28 and OpenSSL 3 for manylinux_2_34.

File: scripts/test_release_pipeline.py [hunk 9]
```diff
@@ -466,6 +478,106 @@ def test_nonofficial_nuget_versions_follow_python_distribution(
...
+        parameter["name"]: parameter
+        for parameter in odbc_template["parameters"]
+    }
+    assert parameters["maxGlibcVersion"]["default"] == "2.34"
+
+    build_script = (_ROOT / "scripts" / "build-odbc-driver-only.sh").read_text(
+        encoding="utf-8"
+    )
+    assert 'MAX_GLIBC_VERSION="${MAX_GLIBC_VERSION:-2.34}"' in build_script
+    assert 'ceiling="GLIBC_$MAX_GLIBC_VERSION"' in build_script
```
[the remark this answers is not in the document]
Comment (gargsaumya) [reply scripts/test_release_pipeline.py]: Valid. Fixed in 3838b6cd. The test now executes the real build script with stubbed Cargo, finalization, and readelf inputs. It proves a 2.28 ceiling accepts GLIBC_2.28 and rejects GLIBC_2.29; with MAX_GLIBC_VERSION unset, the default accepts GLIBC_2.34 and rejects GLIBC_2.35. The /workspace production default is preserved through an overridable WORKSPACE_DIR used only to isolate the test. Validation: 7 focused GLIBC tests passed; complete focused release suite 176 passed with the 3 established Windows ANSI-rendering cases deselected.

**Review by saurabh500 — COMMENTED** [review]
Unattended PR-review sweep, posted automatically.

**Verdict: Clean — no findings. This is not an approval.**

Reviewed the complete 12-file diff at `3838b6cd841af497524f390413858131457924fa` against merge base `c62e8b46a1d02ae8d08531084d23521752834384`, including both architectures, ODBC-disabled paths, shared build/repair callers, the 44-wheel matrix, and prior discussion. I found no additional actionable issues; the previously resolved points are not repeated.

**Blocking:** None.

**Suggestion:** None.

**Nit:** None.

**Verification:** Nine selected structural pipeline tests passed. Removing the x64 explicit image pull and changing its GLIBC ceiling parameter from 2.28 to 2.34 caused the two expected assertion failures; the other seven cases, including ARM64 controls, stayed green. After restoring both changes, the tracked diff was clean and all nine passed again. These checks establish template wiring and test sensitivity, not actual image freshness or binary compatibility. The independent read-only rubber-duck critique reported no significant issues and proposed no corrections.

**Unexercised:** Container/image and wheel builds, Cargo checks, the executable driver-GLIBC regression, wheel-validator execution, real auditwheel repair, ABI/import checks, and live SQL Server paths. Docker was unavailable and ADO work-item access required authentication not available to this run. Required PR validation for this head remains queued (175031); the earlier producer result at `70795fc4` is not counted as validation of this head.

**Review by Theekshna — COMMENTED** [review]
Unattended PR-review sweep (automated, not checked by a human before posting).

Reviewed the full 12-file diff at `3838b6cd841af497524f390413858131457924fa` against merge base `c62e8b46a1d02ae8d08531084d23521752834384`, having verified this checkout's HEAD matches the PR's live `headRefOid` immediately before and after the review. Read all prior discussion first (Copilot bot review threads, the author's own replies/fixes, and an existing automated review under `saurabh500`'s account) before drafting anything new, so none of the already-resolved Copilot findings (GLIBC ceiling ARM64 gap, missing `REPAIR_EXCLUDE_LIBS` coverage, `buildOdbcNative` coupling on both architectures, stale `:latest` image reuse, template header docs, `pyproject.toml` OpenSSL comment) are repeated here — I independently re-verified each fix against the current diff rather than trusting the "fixed" replies at face value.

**Verification performed (not just reading):**
- Ran `scripts/test_release_pipeline.py -k glibc_ceiling` and `scripts/test_repair_glibc_wheels.py`: 15/15 relevant tests pass locally.
- Mutation check on the safety-critical change: inverted the comparison operator in `scripts/build-odbc-driver-only.sh`'s GLIBC-ceiling guard (`!=` → `==`). `test_odbc_build_script_enforces_glibc_ceiling[2.28-2.29-False]` and `[...-2.35-False]` correctly failed, confirming the new test (added in `3838b6cd` to address the earlier Copilot finding that it "only searches script text") actually exercises the guard rather than being vacuous. Restored the file afterward; suite is green again.
- Re-derived the 44-wheel arithmetic myself from `.pipeline/scripts/verify-python-wheels.ps1`'s `Get-ExpectedWheelNames` (5 Python tags × 8 platforms = 40, + 4 `win_arm64` tags excluding cp310 = 44) rather than accepting the PR description's total — it matches, including the 10/10/10/5/9 platform breakdown claimed in the PR body.
- Diffed `.pipeline/OneBranch/stages.yml` for both the x64 and ARM64 arms of the build matrix (the PR edits a duplicated per-architecture table): the `buildOdbcNative` coupling fix and the explicit `docker pull` before the wheel-build step are present and symmetric on both arms, not just the one a reviewer happened to look at first.
- `msodbcsql` parity (`C:\REPOS\Drivers\ODBC_2`): N/A — this PR only touches Python-wheel packaging/build scripts and the ODBC driver's build-time GLIBC-floor check, not TDS/ODBC runtime behavior.
- Divergence-documentation check: N/A — no behavioral divergence from msodbcsql/SqlClient/mssql-jdbc is introduced.
- No AI-slop comments found; the new/updated comments explain *why* (OpenSSL runtime linkage per glibc floor, `AUDITWHEEL_PLAT` collision) rather than restating the diff.

| Severity | Count |
|---|---|
| Blocking | 1 |
| Suggestion | 0 |
| Nit | 0 |

**[Blocking]** `gh pr checks` at this exact head shows the required `mssql-rs Pull request validation` pipeline (ADO build [175031](https://sqlclientdrivers.visualstudio.com/904996cc-6198-4d39-8540-eca72bdf0b7b/_build/results?buildId=175031)) completed with `result: failed`, and the downstream GitHub Action `coverage-report` also fails (`Download Cobertura coverage artifact` step), both confirmed against `pr.sourceSha=3838b6cd841af497524f390413858131457924fa` via the ADO REST API directly (this build is for the reviewed head, not a stale/superseded one). Traced the root cause via the build timeline rather than asserting it: every step in the failing `Build mssql-python macOS (cross-repo)` job up through wheel building (`Build ddbc_bindings and mssql-py-core`, Docker/Colima install) succeeded; only `Start SQL Server (Docker, no TLS certs)` ran ~14 minutes (07:52:15–08:06:03Z) and was canceled, which cascaded the job and the pipeline to `failed` and starved `coverage-report` of its artifact. This PR's diff touches no SQL Server/Docker-e2e infrastructure (confirmed against the 12-file diff), so this reads as a pre-existing macOS Colima/Docker startup flake rather than something introduced here — but I can't fully rule out a interaction, and regardless, the checklist's "Artifact-producing head validation is green; latest test-only head checks are tracked by the PR" was written while this run was still queued/running (per the confirmed 07:20:57Z prior review) and no longer reflects its now-completed, failed state. This is a required check currently blocking merge; needs a human to confirm whether it's a known flake before treating it as safe to proceed, since re-running it is outside this automation's authority.

This review was produced by an unattended scheduled run.

**Comment by Theekshna** [comment]
PR description has mojibake characters

**Review by Theekshna — COMMENTED** [review]
Unattended PR-review sweep (automated, not checked by a human before posting).

This head (`c4afca05`) was pushed since my last review at `3838b6cd`, but `git diff 3838b6cd c4afca05` is **empty** — same tree, only a rebase (parent commit changed). Not re-running the full mutation check / wheel-arithmetic re-derivation from the prior pass against byte-identical content; that verification stands. What changed and needed fresh checking:

- **Prior Blocking finding resolved.** `gh pr checks` at this head shows the ADO `mssql-rs Pull request validation` pipeline (build [175051](https://sqlclientdrivers.visualstudio.com/904996cc-6198-4d39-8540-eca72bdf0b7b/_build/results?buildId=175051)) and every downstream job, including `Build mssql-python macOS (cross-repo)` and `coverage-report`, now **pass**. This confirms my prior read that the earlier failure (`Start SQL Server (Docker, no TLS certs)` canceled after ~14 min) was a transient Colima/Docker flake unrelated to this PR's diff, not a real regression — a clean rebuild of the same tree went green.
- **Review threads**: all 9 threads on this PR (Copilot bot + reassessed) are resolved; no new inline discussion since the last pass.
- **msodbcsql parity** (`C:\REPOS\Drivers\ODBC_2`): N/A — unchanged from last review; this PR is Python-wheel packaging/build-script and GLIBC-floor test coverage only, no TDS/ODBC runtime behavior.
- **Divergence documentation**: N/A — no behavioral divergence from msodbcsql/SqlClient/mssql-jdbc.
- **Verbose slop**: none found; unchanged from last review.
- **Evidence audit**: the 44-wheel arithmetic and the symmetric x64/ARM64 `stages.yml` edits were independently re-derived last pass against this same tree; nothing in the description's evidence table changed.

**Two Nits, both against the PR description (not the diff, so no anchorable line):**

**[Nit]** The "Final-head validation" section still reads `Current source: 3838b6cd841af497524f390413858131457924fa`, but the actual current head is `c4afca0527b627f898cffaa541338021c281624f` (same tree, rebased). Worth updating so the description matches the live head SHA the pipeline validated.

**[Nit]** The description still has mojibake in place of en-dashes (e.g. "glibc 2.28<mojibake>2.33" in the compatibility table, "CPython 3.10<mojibake>3.14" in the validation section) -- I flagged this in a plain comment on 2026-09-14 and it has not been corrected since. Cosmetic only, doesn't affect the pipeline or code, but worth a quick edit for readability.

| Severity | Count |
|---|---|
| Blocking | 0 |
| Suggestion | 0 |
| Nit | 2 |

This review was produced by an unattended scheduled run.

**Review by Vahid-b — COMMENTED** [review]
Unattended PR-review sweep (automated, not checked by a human before posting).

Reviewed the full 12-file diff at `c4afca0527b627f898cffaa541338021c281624f` against merge base `c62e8b46a1d02ae8d08531084d23521752834384` in an isolated detached worktree, after confirming the live repo/author/head immediately before and after. Read all prior discussion first (9 resolved Copilot threads + two earlier automated sweeps) and deliberately did **not** repeat any of it: the ARM64 GLIBC-ceiling gap, `REPAIR_EXCLUDE_LIBS` script coverage, `buildOdbcNative` coupling, stale `:latest` image reuse, template/`pyproject.toml` doc updates, the stale `Current source:` SHA, and the description mojibake are all already raised.

**Verification performed (not just reading):**
- Re-derived the 2.28/2.34 step ordering on **both** architecture arms of `stages.yml`: the 2.28 ODBC build (x64 L327 / ARM64 L471) lands *after* the 2.34 + musl `Stage ... driver` copies and the 2.34 injection, so rebuilding into the shared `odbc-drop` cannot clobber the 2.34/musl drivers. Correct as written.
- Confirmed `mssql-odbc/scripts/finalize-artifact.sh` resolves its output path via `cargo metadata --format-version 1`, which honours the newly-injected `CARGO_TARGET_DIR=/tmp/mssql-odbc-target`. The new isolation env var does not break artifact discovery for any of the three callers (2.28 / 2.34 / musl).
- **Executed** the new GLIBC-ceiling guard in `scripts/build-odbc-driver-only.sh` directly under `sh`, with stubbed `cargo`/`readelf`/`finalize-artifact.sh`, across 7 realistic symbol sets. All correct: `2.28` ceiling accepts `GLIBC_2.2.5 2.17 2.28` and rejects `2.29`/`2.34`; `GLIBC_2.9` is correctly ordered *below* `2.28` by `sort -V`; `GLIBC_PRIVATE` is correctly excluded by the `[0-9.]+` character class and a `GLIBC_PRIVATE`-only input hits the intended `no GLIBC version needs found` fatal path; with `MAX_GLIBC_VERSION` unset the default accepts `2.34` and rejects `2.35`. The guard is sound, including the sub-version and non-numeric edge cases the unit test does not parametrize.
- **Executed** `scripts/repair-glibc-wheels-in-container.sh` against a fake `auditwheel` to check the new `REPAIR_EXCLUDE_LIBS` array plumbing (see Suggestion below for what that turned up). Confirmed the `exclude_args=()` expansion is safe under `set -u` on the bash versions in these images.
- Confirmed `verify-python-wheels.ps1` needs no new `Get-ExpectedOdbcMembers` arm: `manylinux_2_28_*` falls through to the generic `*_x86_64.whl` / `*_aarch64.whl` glibc cases, and `Test-PlatformTag`'s `^manylinux_\d+_\d+_(x86_64|aarch64)$` already accepts the 2.28 tag.
- Reasoned through the "wrong driver silently shipped in a 2.28 wheel" scenario and satisfied myself it is **not** reachable: injection happens *before* `auditwheel repair`, so `auditwheel repair --plat manylinux_2_28` analyses the injected `mssqlodbc.so` too and would reject a 2.34-floor driver. No finding.
- `gh pr checks` at this exact head: all 20 checks pass, including `mssql-rs Pull request validation` (ADO build 175051) and `coverage-report`. The prior sweep's Blocking finding is genuinely resolved.
- `bash -n` clean on both changed shell scripts. Work item `AB#47953` is linked per the PR template.

| Severity | Count |
|---|---|
| Blocking | 0 |
| Suggestion | 1 |
| Nit | 3 |

---

**[Suggestion]** `scripts/repair-glibc-wheels-in-container.sh:18` — `REPAIR_EXCLUDE_LIBS` and `REPAIR_PLAT` are independent env vars, but only one of them has a *safe* default. `EXCLUDE_LIBS="${REPAIR_EXCLUDE_LIBS:-libssl.so.3 libcrypto.so.3}"` keeps the OpenSSL-3 sonames regardless of the plat being repaired, so a `manylinux_2_28` repair that loses its exclusion var degrades silently instead of failing.

I verified this by running the script with a fake `auditwheel`:

| `REPAIR_PLAT` | `REPAIR_EXCLUDE_LIBS` | excludes actually passed | output tag | rc |
|---|---|---|---|---|
| `manylinux_2_28` | `libssl.so.1.1 libcrypto.so.1.1` | `.so.1.1` ✅ | `manylinux_2_28_x86_64` | 0 |
| `manylinux_2_28` | *(unset)* | `.so.3` ❌ | `manylinux_2_28_x86_64` | 0 |
| `manylinux_2_28` | `''` (empty) | `.so.3` ❌ | `manylinux_2_28_x86_64` | 0 |

Note the empty-string row: `${VAR:-default}` treats empty as unset, so a typo'd or blank pipeline variable degrades identically to a missing one. In the real AlmaLinux 8 image the extension links `libssl.so.1.1`, which the `.so.3` exclusions would not match — `auditwheel` would then **graft OpenSSL 1.1 into the wheel**, which is precisely the outcome the header comment says must be avoided ("vendoring the build image's OpenSSL would break server-certificate validation on the target distro" via the compiled-in `OPENSSLDIR`). The wheel still gets a valid `manylinux_2_28` tag, so nothing downstream notices: `verify-python-wheels.ps1` checks names, metadata and ODBC-member presence, not vendored libraries.

This is not a live defect — `stages.yml:363` and `:503` do pass the right sonames on both arms today, and I confirmed that. It is a robustness + test-wiring gap, and it is specifically *not* covered by the regression test added for the earlier Copilot thread: `test_repairs_manylinux_228_with_openssl_11_exclusions` exercises the **script** given the right env, while `test_manylinux_repair_does_not_depend_on_odbc` only asserts step `displayName`s and ordering. Grepping the three test files, `REPAIR_EXCLUDE_LIBS` appears only inside `test_repair_glibc_wheels.py` — nothing asserts the **pipeline** supplies it. Dropping `-e "REPAIR_EXCLUDE_LIBS=..."` from either arm of `stages.yml` would keep the entire suite green and ship vendored OpenSSL.

Two cheap, independent fixes (either helps; both is better):
1. Derive the default from the plat rather than hard-coding `.so.3`, e.g. `case "$PLAT" in *_2_28) default_excludes="libssl.so.1.1 libcrypto.so.1.1";; *) default_excludes="libssl.so.3 libcrypto.so.3";; esac`, or hard-fail when `REPAIR_PLAT` is `manylinux_2_28` and the resolved exclusions still name `.so.3`.
2. Add a `stages.yml` assertion to `test_release_pipeline.py` (parametrized over `Linux_x64` / `Linux_ARM64`, matching the style of `test_manylinux_228_builds_use_isolated_cargo_targets`) that the `Repair glibc-2.28 wheels (Linux {arch})` step's script contains both `REPAIR_PLAT=manylinux_2_28` and `REPAIR_EXCLUDE_LIBS=libssl.so.1.1 libcrypto.so.1.1`.

---

**[Nit]** `.pipeline/OneBranch/stages.yml:286` (and `:430` on ARM64) — the new inline 2.28 wheel-build step drops the "no wheels produced" guard that *both* sibling paths in `build-python-wheels-template.yml` carry (`WHEEL_COUNT=$(find ... -name '*.whl' | wc -l); if [ "$WHEEL_COUNT" -eq 0 ] ...`). It matters slightly more here than for 2.34, because `build-python-wheels-in-container.sh` skips a missing interpreter with `⚠️ Python X not found, skipping...` + `continue` rather than failing, and AlmaLinux 8 is the older base of the two. The zero case is still caught downstream by the repair script's `no bare glibc wheels found to repair`, and a short-but-nonzero set is caught by the publish-stage validator's `Expected 44 wheels, found N` — so this is only about failure locality, not a missed failure. Worth adding the same count check for parity.

**[Nit]** `.pipeline/OneBranch/stages.yml:337` (and `:481` on ARM64) — the new injection step calls `python3` directly, while the sibling 2.34 injection step at `:316` (ARM64 `:460`) guards it with `command -v python3 >/dev/null 2>&1 || { sudo apt-get update && sudo apt-get install -y python3; }`. It works only because both steps live in the same `${{ if and(buildOdbcNative, buildPythonWheels) }}` block and the guarded one always runs first. That is a real but invisible ordering dependency; copying the one-line guard (or hoisting it into its own step) makes the new step self-contained.

**[Nit]** `containers/Dockerfile.PythonBuild.manylinux.2_28.x64` / `.arm64` — these images are now also used to run `auditwheel repair` (`stages.yml:364`, `:504`), but unlike `Dockerfile.PythonBuild.manylinux.x64`, which ends with `RUN which auditwheel && auditwheel --version`, they never assert that precondition at image-build time. The PyPA `manylinux_2_28` base does ship `auditwheel`, so this is correct today; adding the same one-line check keeps the new runtime dependency documented and fails at image build rather than mid-release if a future base sync drops it — exactly the class of surprise the `maturin` verification line just above it guards against.

---

**Not exercised:** container/image builds, the actual `maturin`/`auditwheel` runs, ODBC/wheel binary GLIBC+OpenSSL linkage on real artifacts, and `cargo bfmt`/`bclippy`/`btest`. The Python test suites (`test_release_pipeline.py`, `test_repair_glibc_wheels.py`, `test_verify_python_wheels.py`) could **not** be run in this environment — no real Python interpreter is present (only the Windows Store stub) — so the mutation check reported by the earlier sweep is not independently reproduced here; my script-level conclusions come from executing the shell scripts directly under `bash`, and the test-coverage gap above from reading and grepping the test sources. No SQL Server or live-server paths were touched by this diff.

This PR appears ready for human review, but this is **not** an approval. All findings above are non-blocking.

**Review by shiwanigupta0809 — COMMENTED** [review]
## Summary

This PR adds `manylinux_2_28` / OpenSSL 1.1 wheels alongside the existing `manylinux_2_34` family. I reviewed the 12-file diff at `c4afca0527b627f898cffaa541338021c281624f` against merge base `c62e8b46a1d02ae8d08531084d23521752834384`; all current checks pass, and I found no new significant issues.

## Blocking

None.

## Suggestion

None beyond previously raised review feedback.

## Nit

None beyond previously raised comments, including the PR-description mojibake and stale displayed source SHA.

**Linked work item:** `AB#47953` is present. I could not verify its status/context because Azure DevOps returned `TF401232` (missing item or insufficient permissions).

**Validation context:** Shell syntax checks passed. I could not rerun the targeted Python suites locally because Python was unavailable and `uv` dependency retrieval failed with a TLS handshake error; the corresponding CI validation and coverage checks are green.

**Review by saurabh500 — APPROVED** [review]
